### PR TITLE
Make token-minter more resilient to failures

### PR DESCRIPTION
**Pull request by ironcladlou**
Before this commit, most failure conditions resulted in the token-minter process
exiting, which can cause process restart churn under transient conditions like
the API server not yet being available or a kubeconfig secret not yet being
created.

This commit makes the process always retry failed operations until signalled to
quit. In addition, the process now uses the `log` package so log output contains
timestamps.

### CODE REVIEW

File: token-minter/main.go [hunk 1]
```diff
@@ -112,40 +127,45 @@ func mintToken(ctx context.Context, opts options) metav1.Time {
...
-		panic(err)
+		return metav1.Time{}, fmt.Errorf("failed to get guest kube client: %w", err)
 	}
 
 	// Get the service account
 	var serviceAccount *corev1.ServiceAccount
-	err = wait.PollImmediate(time.Second*5, time.Minute*2, func() (done bool, err error) {
+	timeout, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+	err = wait.PollImmediateUntil(time.Second*5, func() (done bool, err error) {
```
Comment (sjenning) [inline token-minter/main.go:152]: not understanding what this does differently vs the previous `PollImmediate`

Comment (ironcladlou) [reply token-minter/main.go:152]: In this version, the parent context passed to the function can interrupt the poll, whereas in the prior version the poll is not interruptible (e.g. from a process signal).

**Comment by ironcladlou** [comment]
@sjenning 
Test is failing because token-minter is repeatedly failing with an error like this:
```
2022/01/18 00:02:19 error minting token timed out waiting for service account: serviceaccounts "control-plane-operator" not found
2022/01/18 00:02:24 minting token
2022/01/18 00:02:24 error minting token timed out waiting for service account: serviceaccounts "control-plane-operator" not found
2022/01/18 00:02:29 minting token
2022/01/18 00:02:29 error minting token timed out waiting for service account: serviceaccounts "control-plane-operator" not found
2022/01/18 00:02:34 minting token
2022/01/18 00:02:34 error minting token timed out waiting for service account: serviceaccounts "control-plane-operator" not found
```
Looking around it wasn't obvious where that SA is coming from and I don't see it in any of the artifact dumps... ideas?

**Comment by ironcladlou** [comment]
Is it because that SA exists in the guest cluster (whose resources we don't dump?) Still not sure where the SA is created in the first place

**Comment by ironcladlou** [comment]
Looks like the nodes never joined. Not the first time I've seen this in PRs lately...

/retest

File: token-minter/main.go [hunk 2]
```diff
@@ -65,106 +65,113 @@ func main() {
...
 		select {
 		case <-time.After(renewDuration):
-			expirationTimestamp = mintToken(ctx, opts)
+			log.Println("minting token")
+			expirationTimestamp, err := mintToken(ctx, opts)
+			if err != nil {
+				log.Println("error minting token, will retry in", ErrorRetryPeriod, err)
+				renewDuration = ErrorRetryPeriod
+			} else {
+				log.Println("renew delay set for", renewDuration.String())
```
Comment (alvaroaleman) [inline token-minter/main.go:86]: Doesn't this log need to be after actually updating `renewDuration` which currently happens afterwards?

**Comment by ironcladlou** [comment]
The refactor introduced a subtle bug whereby the polling failure for the SA caused the reconcile to fail rather than proceed to the create-if-not-exist path. So the SAs and tokens weren't being created, causing CAPI provider to break.

It occurred to me with the new reconcile model introduced here, there's no need for internal polling/retries within `mintToken()` itself — that's handled now on the outside in the main control/reconcile loop.

So, to make it all behave and also easier to understand, `mintToken()` now just does all its steps in sequence, and if it encounters any failures along the way, it returns an error immediately and lets the outside loop deal with retries.

I also wired in specific timeout contexts for the API calls to ensure they will fail and retry within a reasonable time.

File: token-minter/main.go [hunk 3]
```diff
@@ -65,106 +65,113 @@ func main() {
...
 	if opts.oneshot {
-		return
+		_, err := mintToken(ctx, opts)
+		if err != nil {
+			log.Fatalln(err)
+		}
+		os.Exit(0)
 	}
 
+	renewDuration := 1 * time.Second
```
Comment (sjenning) [inline token-minter/main.go:76]: why set to 1s?  why not just `var renewDuration time.Duration` and move it inside the `for`?

Comment (ironcladlou) [reply token-minter/main.go:76]: I guess it could be the zero value (`int64`) and it would work the same

File: token-minter/main.go (hunk 2 above)
Comment (ironcladlou) [reply token-minter/main.go:86]: Oops! Yes!

File: token-minter/main.go (hunk 3 above)
Comment (ironcladlou) [reply token-minter/main.go]: Fixed

File: token-minter/main.go (hunk 2 above)
Comment (ironcladlou) [reply token-minter/main.go]: Fixed

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**